### ISSUE TITLE
Get Android port ready for 64-bits requirements

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#600]
Google sent me an e-mail that contains the text here:

https://android-developers.googleblog.com/2019/01/get-your-apps-ready-for-64-bit.html

I think it's just adding the **`arm64-v8a`** and **`x86_64`** targets will work but I haven't tested. Note armeabi and mips are dropped for recent ndk.

https://developer.android.com/ndk/guides/abis

Testing here, on `makestandalones.sh` . Platform has to be updated to 21 (from 14) and arch x86_64 and arm64 is added

```
...
PLATFORM=android-21

# standalone toolchains cannot share same directory
export NDK_STANDALONE=$NDK_HOME/platforms/$PLATFORM

# if you don't pass force, the directory is not rewritten
# verbose is needed to see why making the toolchain fails
for arch in arm x86 x86_64 arm64 mips
...

```

Then copied `x86` folder and renamed to `x86_64`, fixed ndkenv, and copied `armeabi-v7a` and renamed to `arm64-v8a`, and also updated it's ndkenv.

x86_64 ndkenv changes
```
export NDK_PLATFORM_ROOT=$NDK_STANDALONE/x86_64
export NDK_PLATFORM_NAME=x86_64
export NDK_HOST_NAME=x86_64-linux-android
```

arm64 ndkenv changes
```
export NDK_PLATFORM_ROOT=$NDK_STANDALONE/arm64
export NDK_PLATFORM_NAME=arm64-v8a
export NDK_HOST_NAME=aarch64-linux-android
```

and both
```
export CC=$NDK_PLATFORM_ROOT/bin/${NDK_HOST_NAME}-gcc
export CXX=$NDK_PLATFORM_ROOT/bin/${NDK_HOST_NAME}-g++ 
```

Then on buildall.sh did:

```
# android-21 is the minimum we can go with current Android SDK and 64-bits
PLATFORM=android-21

# standalone toolchains cannot share same directory
export NDK_STANDALONE=$NDK_HOME/platforms/$PLATFORM

for arch in armeabi armeabi-v7a x86 x86_64 arm64-v8a mips 
```

Lua will fail for platform-21 because bsd_signal is removed. 

`lua.c:function laction: error: undefined reference to 'bsd_signal'`

[There is a fix here](https://stackoverflow.com/questions/36746904/android-linker-undefined-reference-to-bsd-signal)

Now for actually building. I skipped lua for now. x86_64 seems to build fine. For the arm64, I seem to be getting an annoying error:

```
cd builds/unix; /bin/sh ./configure  '--host=aarch64-linux-android' '--prefix=/home/USERNAME/git/ags/Android/nativelibs/arm64-v8a' '--without-zlib' '--disable-shared'
checking build system type... x86_64-unknown-linux-gnu
checking host system type... aarch64-unknown-linux-android
checking for aarch64-linux-android-gcc... /home/USERNAME/Android/Sdk/ndk-bundle/android-ndk-r16b//platforms/android-21/arm64/bin/aarch64-linux-android-gcc
checking whether the C compiler works... no
configure: error: in `/home/USERNAME/git/ags/Android/buildlibs/arm64-v8a/freetype-2.4.12/builds/unix':
configure: error: C compiler cannot create executables

```

[changes_Android64.zip](https://github.com/adventuregamestudio/ags/files/2823958/changes_Android64.zip)

**Comment by ivan-mogilko** [comment]
<pre>
checking whether the C compiler works... no
...
configure: error: C compiler cannot create executables
</pre>

I think this may mean that a necessary tool is not installed or not found. I usually see this error when build Android port after cleaning up enviroment variables (ANDROID_HOME etc) and have to set them up again.

**Comment by morganwillcock** [comment]
Just a guess, but `//` normally means a variable was expanded to nothing, unless this directory structure is correct:
```
/home/USERNAME/Android/Sdk/ndk-bundle/android-ndk-r16b//platforms/android-21/arm64/bin/aarch64-linux-android-gcc
                                                      ^^
```

**Comment by ivan-mogilko** [comment]
@ericoporto , could you comment on this, is this still an issue, and if so, was there any progress in building port on your side?

**Comment by ericoporto** [comment]
I haven't looked into again since the first time in February. I think right now Google Play is already blocking uploading builds that don't comply with 64bits requirement. Searching around apparently some new text on the subject has surfaced so I will link some more info below.

There are four ABIs that can be built on current Android tools:

- **`armeabi-v7a`**– ARM-based CPUs with hardware floating-point operations and multiple CPU (SMP) devices.

- **`arm64-v8a`** – CPUs based on the 64-bit ARMv8 architecture.

- **`x86`** – CPUs that support the x86 (or IA-32) instruction set.

- **`x86_64`** – CPUs that support the 64-bit x86 (also referred to as x64 and AMD64) instruction set.

So ndk-build would need the following on it's .mk file:

`APP_ABI := armeabi-v7a arm64-v8a x86 x86_64`

The below text from Google has very interesting info:

https://developer.android.com/distribute/best-practices/develop/64-bit

From here:

https://www.codeproject.com/Articles/1288668/Making-sure-your-Android-game-is-ready-for-64-bi-2

> For Android apps, there are 2 settings that you can set to specify your API levels:
>
> - Minimum SDK Level - The lowest API level that your app supports. Any phone that has a lower API version can’t run your app. This setting also limits the features that are available to your app to the ones that are available for your minimum SDK Level
>
> - Target SDK Level - The API level that your app is intended for. Anything below that will run Android’s compatibility code to provide a similar experience for phones that don’t support the Target SDK Level.
>
> From the perspective of having a 64-Bit ready app, we would need to ensure that our Target SDK Level must be at least 21 to allow Android Studio to support a 64-bit architecture. Going forward, though, this won’t be enough.
>
> As mentioned, if you are adding a new app to the Play Store, or updating an existing app, you must target API level 26 (Android 8.0) or higher. And as Google outlined in [a recent blog post](https://android-developers.googleblog.com/2019/02/expanding-target-api-level-requirements.html), new applications will be required to have a target API level of 28 (Android 9.0) beginning in August 2019.
>
> If you’re making updates to an existing application, you’ll have to target API level 28 beginning in November 2019. Google has compiled a [list of important](https://developer.android.com/distribute/best-practices/develop/target-sdk)

And lastly, this one shows how to generate a 64 bit toolchain, which I still don't know if it's  needed.

https://web.archive.org/web/20190928124535/https://www.tal.org/tutorials/openssl_qt_android

**Comment by ericoporto** [comment]
Got builds working here: https://github.com/ericoporto/ags/tree/android-64bit
The CI is going here with binary builds: https://cirrus-ci.com/github/ericoporto/ags/android-64bit

What's not working: sound, you can hear things but there's a lot of noise.

**Comment by ericoporto** [comment]
Got a new email from Google Today.

>Hello Google Play  Developer,
> 
> This is a reminder that starting November 1, 2019, updates to apps and games on Google Play will be [required to target Android 9 (API level 28) or higher](https://support.google.com/googleplay/android-developer/answer/113469#targetsdk). After this date, the Play Console will prevent you from submitting new APKs with a targetSdkVersion less than 28.
> 
> Configuring your app to target a recent API level ensures that users benefit from significant security and performance improvements, while still allowing your app to run on older Android versions (down to the minSdkVersion).
> 
> Action required
> 
> Please ensure that your apps are configured to target at least Android 9 (API level 28) by November 1, 2019. For technical advice on how to change your app's target API level to meet these requirements, refer to the [migration guide](https://developer.android.com/distribute/best-practices/develop/target-sdk).

**Comment by ericoporto** [comment]
Google has hidden my Adventure Game Studio games from Google Play search results (they can still be found through Google). I have no idea if it's related to this, but since the port as is, is working on Android 64-bits, I will try later to update them to use it so I can comply with the policy changes.

**Comment by ericoporto** [comment]
I decided to iterate Sonneveld's work.

I rebased and got it working on Linux too:
https://github.com/ericoporto/ags/tree/ags3--android-studio

I would like to also make it working again on Windows and then evaluate what to keep and what not to keep.

**Comment by ericoporto** [comment]
Just note that the branch with opengles2 render alone in the end of last comment also includes the current OpenGL render in the same engine, you run the using either `--gfxdriver ogles2` or `--gfxdriver ogl`. It's meant to test the renderer logic for cameras that are different from when the android studio branch was done. The new renderer depends on glm. In the long future glm is useful for a vulkan renderer too.

**Comment by ivan-mogilko** [comment]
I think this was resolved by #1190 , @ericoporto could you confirm this?
